lib/api/apiUtils/object/coldStorage.js

```diff
@@ -0,0 +1,107 @@
+const ObjectMDArchive = require('arsenal').models.ObjectMDArchive;
+const errors = require('arsenal').errors;
+const { config } = require('../../../Config');
+const { locationConstraints } = config;
+
+/**
+ * Check restore status.
+ *
+ * @param {ObjectMD} objectMD - object metadata
+ * @return {undefined}
+ */
+function _checkRestoreStatus(objectMD) {
+    if (objectMD.archive?.restoreRequestedAt && !objectMD.archive?.restoreCompletedAt) {
+        // check if archive.restoreRequestAt exists and archive.restoreCompletedAt not yet exists
+        return 'RestorePending';
+    } else if (objectMD.archive?.restoreCompletedAt
+        && new Date(objectMD.archive?.restoreWillExpireAt) >= new Date(Date.now())) {
+        //  check if archive.restoreWillExpireAt > current time
+        return 'Restored';
+    } else if (objectMD.archive?.restoreCompletedAt
+        && new Date(objectMD.archive?.restoreWillExpireAt) < new Date(Date.now())) {
+        return errors.InvalidObjectState;
+    } else if (objectMD.archive?.archiveInfo) {
+        return 'Archived';
+    } else {
+        return errors.InvalidObjectState;
+    }
+}
+
+/**
+ * Check if restore can be done.
+ *
+ * @param {BucketInfo} bucketMD - bucket instance
+ * @param {ObjectMD} objectMD - object metadata
+ * @return {Error|boolean} - true if the object is already restored
+ */
+function _validateStartRestore(bucketMD, objectMD) {
+    const isLocationCold = locationConstraints[bucketMD.getLocationConstraint()]?.cold;
+    if (isLocationCold === undefined || isLocationCold === false) {
+        // either location cold flag not exists or cold flag is explicit false
+        return errors.InvalidObjectState;
+    }
+    const restoreStatus = _checkRestoreStatus(objectMD);
+    if (restoreStatus instanceof Error) {
+        return errors.InvalidObjectState;
+    }
+    if (restoreStatus === 'RestorePending') {
+        return errors.RestoreAlreadyInProgress;
+    } else {
+        return restoreStatus === 'Restored';
+    }
+}
+
+/**
+ * update restore expiration date.
+ *
+ * @param {BucketInfo} bucketMD - bucket instance
+ * @param {ObjectMD} objectMD - objectMD instance
+ * @param {object} restoreParam - bucket name
+ * @param {object} log - werelogs logger
+ * @return {undefined|Error} internal error if object MD is not valid
+ *
+ */
+function _updateRestoreInfo(bucketMD, objectMD, restoreParam, log) {
+    try {
+        // eslint-disable-next-line no-param-reassign
+        objectMD.archive.restoreRequestedAt = new Date();
+        // eslint-disable-next-line no-param-reassign
+        objectMD.archive.restoreRequestedDays = restoreParam.days;
+        if (!ObjectMDArchive.isValid(objectMD.archive)) {
+            throw new Error('archive object is not valid');
+        }
+    } catch (e) {
+        log.debug(e.message, { md: objectMD.archive });
+        return errors.InternalError;
```

Comment:
I'm not sure that `InternalError` is the right kind of error (we use it too much to the point of not being very helpful).

Reply:
Maybe the `BadMetadataDocument` can be used here?

Reply:
At least `InternalError.customizeDescription(...)` would be more useful.

Reply:
done: https://github.com/scality/cloudserver/pull/4507/commits/a7016fef2534bc31a385438e9cd86745b1eee72e